src/home/homework-blob.c

```diff
@@ -0,0 +1,236 @@
+                RecurseDirEvent event,
+                const char *path,
+                int dfd,
+                int fd, /* unset! */
+                const struct dirent *de,
+                const struct statx *sx,
+                void *userdata) {
+
+        _cleanup_close_ int src = -EBADF;
+        struct blob_copy_data *d = ASSERT_PTR(userdata);
+        int r;
+
+        if (event == RECURSE_DIR_ENTER) {
+                log_warning("Entry %s in blob directory is a directory. Skipping.", de->d_name);
+                return RECURSE_DIR_SKIP_ENTRY;
+        }
+        if (event != RECURSE_DIR_ENTRY)
+                return RECURSE_DIR_CONTINUE;
+
+        if (!S_ISREG(sx->stx_mode)) {
+                log_warning("Entry %s in blob directory is not a regular file. Skipping.", de->d_name);
+                return RECURSE_DIR_CONTINUE;
+        }
+
+        src = openat(dfd, de->d_name, O_RDONLY|O_CLOEXEC|O_NOCTTY|O_NOFOLLOW);
+        if (src < 0)
+                return log_debug_errno(errno, "Failed to open %s in src blob: %m", path);
+
+        r = copy_one_blob(src, d->dest_dfd, de->d_name, &d->total_size, d->uid, d->manifest);
+        if (r == -EOVERFLOW)
+                return RECURSE_DIR_LEAVE_DIRECTORY;
+        if (r < 0)
+                return r;
+        return RECURSE_DIR_CONTINUE;
+}
+
+static int replace_blob_at(
+                int src_dfd,
+                const char *src_name,
+                int dest_dfd,
+                const char *dest_name,
+                Hashmap *manifest,
+                mode_t mode,
+                uid_t uid) {
+        _cleanup_free_ char *fn = NULL;
+        _cleanup_close_ int src_fd = -EBADF, dest_fd = -EBADF;
+        struct blob_copy_data userdata;
+        int r;
+
+        assert(src_dfd >= 0);
+        assert(src_name);
+        assert(dest_dfd >= 0);
+        assert(dest_name);
+
+        src_fd = openat(src_dfd, src_name, O_RDONLY|O_DIRECTORY|O_CLOEXEC|O_NOFOLLOW);
+        if (src_fd < 0) {
+                if (errno == ENOENT)
+                        return 0;
+                return log_debug_errno(errno, "Failed to open src blob dir: %m");
+        }
+
+        r = tempfn_random(dest_name, NULL, &fn);
+        if (r < 0)
+                return r;
+
+        dest_fd = open_mkdir_at(dest_dfd, fn, O_EXCL|O_CLOEXEC, mode);
+        if (dest_fd < 0)
+                return log_debug_errno(dest_fd, "Failed to create/open dest blob dir: %m");
+
+        /* Note: We do it this way instead of just using FOREACH_DIRENT so that we
+         * walk the dirents in alphabetical order and thus behave deterministically
+         * w.r.t. what happens if the dir hits its size quota */
+        userdata = (struct blob_copy_data) {
+                .dest_dfd = dest_fd,
+                .total_size = 0,
+                .uid = uid,
+                .manifest = manifest,
+        };
+        r = recurse_dir_at(src_fd, ".", STATX_TYPE|STATX_SIZE, UINT_MAX, RECURSE_DIR_SORT,
```

Comment:
given you don't actually intend to descend into any subdirs here, you could also use readdir_all() with RECURSE_DIR_ENSURE_TYPE

Reply:
also STATX_SIZE you don't actually use, do you?

Reply:
I didn't use RECURSE_DIR_ENSURE_TYPE since I stat in `copy_one_blob` now to enforce the type of the FD. But done.